synapse/handlers/room_summary.py

```diff
@@ -803,8 +803,10 @@ async def _summarize_remote_room_hierarchy(
             if "room_id" in c and isinstance(c["room_id"], str)
         }
 
+        room_response = dict(room_response)
+        children_state = room_response.pop("children_state", ())
         return (
-            _RoomEntry(room_id, room_response, room_response.pop("children_state", ())),
```

Comment:
room_response is a JsonDict, which is why .pop() is allowed.

This bug feels like something that mypy could have prevented if we had a JsonMapping type. Except we can't just define it as a `Mapping`, since Mappings aren't necessarily JSON-serializable.

Reply:
Maybe https://docs.python.org/3.10/library/types.html#types.MappingProxyType, but we'd have to register something to make those json encodable. NewType probably the best way forward, short of some kind of `const/mutable` semantics in the type system.